[REFACTOR] 공용 컴포넌트 분리 (#285)

ParkSY0919 opened the pull request:
# 🔥*Pull requests*

## ⛳ **작업한 브랜치**
- #285 

<br>

## 👷 **작업한 내용**
<!-- 작업한 내용을 적어주세요. -->
#### `코스 등록`, `일정 등록` 화면에서 사용되는 공용 컴포넌트 분리
  - [x] DRTextField
#### 코스등록
  - [x] 등록 사진 삭제 시 label에 표시된 사진 개수와 동기화

<br>

## 📌 DRTextField
> 데이트로드에서 사용되는 TextField 컴포넌트

### 1️⃣ 제작 이유
<img width="300" alt="스크린샷 2025-03-15 오전 1 05 52" src="https://github.com/user-attachments/assets/4eefff83-877f-4e1c-b50f-2dba7c09bdcc" />

  - 현재 데이트로드에서 사용되는 텍스트 필드는 위 사진과같은 형태로 placeholder의 text 차이와 rightView의 타입차이만 있기에 이를 모두 관통할 수 있는 컴포넌트를 만들어 코드의 재활용성을 높여보고자 만들게 되었습니다.

### 2️⃣ DRTextField 설명

#### DRTextFieldType
  - DRTextFieldStyle의 style을 사용되는 case 별로 나눠 미리 정의한 enum으로 이는 DRTextField를 생성할 때에 사용됩니다.

#### DRTextFieldStyle
  - textField가 어떠한 형태를 이루는지 case 별로 정의한 enum으로 기본 매개변수로는 placeHolder와 cornerRadius 등이 있습니다.

#### RightIconStyle
  - textField rightView에서 사용되는 Icon의 스타일을 정의한 enum 입니다.

#### Flow
  - UITextField를 활용하여 만들어진 `DRTextField`는 데이트로드 어플 내에서 사용되는 textField의 각 case들이 정리된 `DRTextFieldType`를 생성할 때에 매개변수로 받아 정의된 style 값을 활용하여 TextField를 그려냅니다.

#### Init

  ```swift
  init(type: DRTextFieldType) {
      self.isNumberPad = type == DRTextFieldType.AddCourseSchedule(.totalPrice) ? true : false
      self.isRightButton = type == DRTextFieldType.profile ? true : false
      
      self.type = type.textFieldStyle
      self.leftPadding = type.textFieldStyle.leftPadding
      self.rightPadding = type.textFieldStyle.rightPadding
      super.init(frame: .zero)
      
      setupTextField()
      setupRightView()
  }
  ```

  - `self.isNumberPad` : 텍스트필드 키보드 타입 핸들링 변수
    - 현재는 DRTextFieldType에 ‘총 금액’을 입력하는 경우에만 넘버패드를 사용하기에 위와 같이 구성하였습니다.
  - `self.isRightButton` : 텍스트필드 rightView 핸들링 변수
    - 현재는 ‘프로필 닉네임’ 관련 작성 시에만 rightView에 버튼이 위치하기에 위처럼 구성하였습니다.
  - `self.leftPadding, self.rightPadding` : 텍스트필드의 좌, 우 패딩

#### setupTextField(), setupRightView()

  - 데이트로드에서 사용되는 텍스트필드의 공통 속성 설정과 각 case별 세부로 적용되는 속성 및 rightView를 설정하는 함수입니다.

#### 기타 함수

  - calculateRightPadding()
      - 우측 패딩 계산 함수
  - func textRect(forBounds bounds: CGRect)
      - 기본 상태일 때에 text가 표시될 수 있는 영역에 대한 함수
  - func editingRect(forBounds bounds: CGRect)
      - 편집 중일 때에 text가 표시될 수 있는 영역에 대한 함수
  - func leftViewRect(forBounds bounds: CGRect)
      - textField leftView가 위치할 영역 (현재의 경우 여백)
  - func rightViewRect(forBounds bounds: CGRect)
      - textField rightView가 위치할 영역 (현재의 경우 아이콘 or 버튼)

### 3️⃣ case 추가하여 사용하는 방법

  - DRTextFieldStyle에 맞게 case를 추가한 이후 DRTextFieldType에 해당 텍스트필드를 사용할 곳의 case를 추가하여 DRTextField.DRTextFieldStyle의 속성을 미리 정의해준다면, 이후 아래와 같이 호출만 진행하면 됩니다.
 
      ```swift
      private let nicknameTextfield = DRTextField(type: .profile)
      ```
    

### 4️⃣ DRTextField 프로퍼티로 추가한 이후 해야할 것

  - textField를 제대로 사용하기 위해서는 프로퍼티로 추가한 이후 아래 3단계를 진행해주어야 합니다.
      1. Hierarchy에 추가하기
      2. Layout 잡아주기
      3. 활용되는 font 적용해주기

### 5️⃣ DRTextField의 이점

  - **일관된 UI/UX 제공**
      - 앱 전체에서 동일한 스타일과 동작 유지
  - **코드 재사용성 향상**
      - 반복되는 텍스트 필드 구현 없이 재사용 가능
  - **유지보수 용이성**
      - 한 곳에서 수정하면 모든 필드에 적용됨
  - **타입 안전성 보장**
      - `enum`을 활용해 명확한 타입 관리 가능
  - **확장성**
      - 새로운 스타일을 쉽게 추가 및 관리 가능
  - **디자인 시스템 통합**
      - 디자인 시스템을 코드에 반영하여 협업 강화
  - **테스트 용이성**
      - 독립적으로 테스트 가능하여 품질 관리 쉬움
  - **설정 오류 방지**
      - 타입에 따라 속성이 자동 설정되어 실수 방지

<br>

### 🩹Fix: 등록 사진 삭제 시 label에 표시된 사진 개수와 동기화

|    기능    |   스크린샷   |
| :-------------: | :----------: |
| GIF | <img src = "https://github.com/user-attachments/assets/9e24f2e0-e2fc-409c-954c-eb2035ac206f" width ="220">|

<br>


## 📟 관련 이슈
- Resolved: #285

ParkSY0919 committed:
[Feat] DRTextField 구현 (#285)

ParkSY0919 committed:
[Refactor] '코스 등록 뷰1' DRTextField 적용 (#285)

ParkSY0919 committed:
[Feat] DRCommonButton 구현 (#285)

ParkSY0919 committed:
[Fix] DRCommonButton 속성 설정 수정 (#285)

ParkSY0919 committed:
[Fix] DRCommonButton 속성 설정 관련 코드 수정 등 (#285)

ParkSY0919 committed:
[Refactor] '코스 등록 뷰1' DRCommonButton 적용 (#285)

ParkSY0919 committed:
[Fix] 하단 버튼 레이아웃 조정 (#285)

ParkSY0919 committed:
[Feat] DRCommonButton convenience init 추가 (#285)

ParkSY0919 committed:
[Feat] DRTextFieldType  생성 (#285)

ParkSY0919 committed:
[Feat] DRTextField 적용 (#285)

ParkSY0919 committed:
[Feat] DRTextField profile case 추가 (#285)

ParkSY0919 committed:
[Fix] 사용하지 않는 코드 삭제 및 수정 (#285)

ParkSY0919 committed:
[Del] DRTextField 등 적용 이후 미사용 코드 삭제 (#285)

cirtuare commented on DATEROAD-iOS/DATEROAD-iOS/Presentation/AddCourse/Views/AddCourseFirst/AddFirstView.swift:
```diff
@@ -169,7 +111,7 @@ final class AddFirstView: BaseView {
         }
         
-        [dateNameTextField, visitDateContainer, dateStartAtContainer].forEach { view in
+        [ dateStartAtTextField].forEach { view in
```
앞에 공백 하나 들어갔습니당 .. !!

cirtuare reviewed (approved):
고생하셨습니당 ~~~~~ 💪

ParkSY0919 committed:
[Fix] DRComponent 적용 이전 코드로 복구 (#285)

어머니 코드 merge 시 conflict 방지

ParkSY0919 committed:
[Fix] fix Conflict to Merge

# Conflicts:
#	DATEROAD-iOS/DATEROAD-iOS.xcodeproj/project.pbxproj

ParkSY0919 committed:
[Del] 사용하지 않는 파일 삭제 (#285)

ParkSY0919 committed:
[Fix] 코스 등록 버튼 레이아웃 수정

ParkSY0919 committed:
[Fix] 코스 등록 버튼 레이아웃 수정 (#285)

ParkSY0919 committed:
[Fix] 코스&일정 등록 레이아웃 조정

ParkSY0919 committed:
[Fix] 코스&일정 등록 레이아웃 조정 (#285)

ParkSY0919 committed:
[Feat] 코스&장소 등록 DRTextField 적용 (#285)

ParkSY0919 committed:
[Feat] 코스&장소 등록 DRTextField 적용 (#285)

ParkSY0919 committed:
[Fix] textField font 적용 (#285)

ParkSY0919 committed:
[Fix] textField font 적용 (#285)

ParkSY0919 committed:
[Del] 사용하지 않는 코드 삭제 (#285)

ParkSY0919 committed:
[Del] 사용하지 않는 코드 삭제 (#285)

ParkSY0919 committed:
[Fix] fix Conflict to Pull

# Conflicts:
#	DATEROAD-iOS/DATEROAD-iOS/Presentation/AddSchedule/Views/AddScheduleFirst/AddScheduleFirstViewController.swift
#	DATEROAD-iOS/DATEROAD-iOS/Presentation/AddSchedule/Views/AddScheduleFirst/InAddScheduleFirstView.swift

ParkSY0919 committed:
[Fix] 선택된 이미지 개수 text와 동기화 (#285)

ParkSY0919 committed:
[Fix] BottomPageControllView index label로 변경 (#285)

ParkSY0919 committed:
[Chore] 주석 추가 (#285)

seuriseuljjeok commented on DATEROAD-iOS/DATEROAD-iOS/Global/UIComponents/DRTextField.swift:
```diff
@@ -0,0 +1,241 @@
+        case totalPrice
+    }
+    
+    case AddCourseSchedule(AddCourseScheduleType)
```
나머지 case는 소문자니까 통일해주면 더 좋을 것 같습니다

seuriseuljjeok commented on DATEROAD-iOS/DATEROAD-iOS/Global/UIComponents/DRTextField.swift:
```diff
@@ -0,0 +1,241 @@
+
+import UIKit
+
+enum DRTextFieldType: Equatable {
```
전체적으로 줄바꿈해주세요!

seuriseuljjeok replied:
얘도 따로 파일 분리하는 게 더 나을 것 같습니다

seuriseuljjeok commented on DATEROAD-iOS/DATEROAD-iOS/Global/UIComponents/DRTextField.swift:
```diff
@@ -0,0 +1,241 @@
+    private func setupRightView() {
+        if isRightButton {
+            let button = UIButton()
+            button.setTitle(StringLiterals.Profile.doubleCheck, for: .normal)
```
DRTextButton 적용해서 통일하면 더 좋을 것 같습니다

seuriseuljjeok commented on DATEROAD-iOS/DATEROAD-iOS/Global/UIComponents/DRTextField.swift:
```diff
@@ -0,0 +1,241 @@
+    private let rightPadding: CGFloat
+    
+    init(type: DRTextFieldType) {
+        self.isNumberPad = type == DRTextFieldType.AddCourseSchedule(.totalPrice) ? true : false
```
`self.isNumberPad = type == DRTextFieldType.AddCourseSchedule(.totalPrice)` 이렇게 축약해도 될 것 같습니다

seuriseuljjeok commented on DATEROAD-iOS/DATEROAD-iOS/Global/UIComponents/DRTextField.swift:
```diff
@@ -0,0 +1,241 @@
+
+final class DRTextField: UITextField {
+    
+    enum DRTextFieldStyle {
```
enum을 final class 안에 넣어둔 이유는 그때 얘기 나눴던 것 때문에 그런 것 같은데, 사실 지금 상황에서는 따로 파일을 분리하는 게 더 나을 것 같아요 어차피 참조를 하는 클래스에서만 사용하기 때문에 메모리 부분에서 큰 차이는 없을 것 같습니다
오히려 DRTextField와 여러 enum들이 한 파일에 있어서 가독성 측면에서 많이 떨어지는 것 같아요!

seuriseuljjeok commented on DATEROAD-iOS/DATEROAD-iOS/Global/UIComponents/DRTextField.swift:
```diff
@@ -0,0 +1,241 @@
+    //텍스트필드 속성 설정
+    private func setupTextField() {
+        setPlaceholder(placeholder: type.placeholderText, fontColor: UIColor(resource: .gray300), font: .suit(.body_semi_13))
+        self.autocorrectionType = .no
```
do,,,,

seuriseuljjeok commented on DATEROAD-iOS/DATEROAD-iOS/Presentation/AddCourse/Views/AddCourseFirst/AddFirstView.swift:
```diff
@@ -64,23 +50,15 @@ final class AddFirstView: BaseView {
-        visitDateContainer.addSubviews(visitDateLabel, visitDateImage)
-        
-        dateStartAtContainer.addSubviews(dateStartTimeLabel, dateStartTimeImage)
+            visitDateTextField,
```
줄바꿈 3개부터!

seuriseuljjeok commented on DATEROAD-iOS/DATEROAD-iOS/Global/UIComponents/DRTextField.swift:
```diff
@@ -0,0 +1,241 @@
+final class DRTextField: UITextField {
+    
+    enum DRTextFieldStyle {
+        case basic(placeholder: String, cornerRadius: CGFloat, alignment: NSTextAlignment = .left)
```
폰트, bgcolor, textcolor 등을 속성으로 추가하지 않은 이유가 있나요?

ParkSY0919 replied:
bgColor와 fontColor는 통일된 색상이기에 배제하였고, font는 추가하였습니다!

seuriseuljjeok commented on DATEROAD-iOS/DATEROAD-iOS/Presentation/AddCourse/Views/AddCourseThird/AddThirdView.swift:
```diff
@@ -91,7 +91,7 @@ final class AddThirdView: BaseView {
                                    font: .suit(.body_bold_17))
         
-        [contentTextView, priceTextField].forEach {
+        [contentTextView].forEach {
```
forEach 사용하지 않아도 될 것 같습니다

seuriseuljjeok commented on DATEROAD-iOS/DATEROAD-iOS/Presentation/AddSchedule/Views/AddScheduleFirst/AddScheduleFirstView.swift:
```diff
@@ -78,7 +78,7 @@ extension AddScheduleFirstView {
         visitDateErrorLabel.isHidden = isPassValid
-        inAddScheduleFirstView.visitDateContainer.do {
+        inAddScheduleFirstView.visitDateTextField.do {
             $0.layer.borderWidth = isPassValid ? 0 : 1
```
속성 하나만 업데이트 해주는거면 do 사용하지 않아도 될 듯요!

seuriseuljjeok reviewed (approved):
매서운 리팩 수고하셨습니다! 전반적으로 줄바꿈이나 컨벤션 확인 부탁드릴게요!

ParkSY0919 committed:
[Fix] 코드리뷰 적용 및 ProfileView DRTextField 적용 (#285)

ParkSY0919 committed:
[Fix] 코드리뷰 적용 (#285)

DRTextField 관련 enum 분리

seuriseuljjeok commented on DATEROAD-iOS/DATEROAD-iOS/Global/Utils/Enums/DRTextFieldType.swift:
```diff
@@ -0,0 +1,181 @@
+
+import UIKit
+
+enum DRTextFieldType: Equatable {
```
슨생님 enum 내부 줄바꿈 매섭게 부탁드림요

ParkSY0919 replied:
바로 꿀밤줘버렸습니다!!

ParkSY0919 committed:
[Fix] codeConvention 적용 (#285)
